### ISSUE TITLE
Transparent reconnects

### ISSUE THREAD

**Issue by kousu** [pudo/dataset#88]
SQLAlchemy [transparently supports connection pooling](http://docs.sqlalchemy.org/en/latest/core/pooling.html), but only if you write your code something like this:

```
engine = create_engine() #global

def handler():
    conn = engine.connect()
    meta = MetaData()
    meta.reflect(conn)
    return HTMLize(conn.execute(select([meta.tables.myawesometable])))
```

I really want to use Dataset though, because you've already done almost all of what I care about. Analogous to the above, I would like to write
```
db = dataset.Database()

def handler():
    try:
        return db['myawesometable'].all()
    except SQLNotConnected:
        return 503
```

However [this line](https://github.com/pudo/dataset/blob/a797e9035720c24da712fb565a1ce6aa4feeac37/dataset/persistence/database.py#L56) is giving me trouble, because I cannot guarantee my database always existing, but when it is up I want it to be responsive.

Once solution I tried was
```
def handler():
    try:
       db = dataset.Database()
    except SQLNotConnected:
        return 503
    
    return db['myawesometable'].all()
```
but this makes it impossible for SQLAlchemy to do any pooling (and, in the case of sqlite://:memory:, actually talks to a different database), because `Database.__init__` calls `create_engine()` itself.

I want to treat `Database` such that an instance is an actual reference to a database, not a reference to _a connection to_ a database.

By tracing down [through dataset](https://github.com/pudo/dataset/blob/a797e9035720c24da712fb565a1ce6aa4feeac37/dataset/persistence/database.py#L65) and into the [bowels of SQLAlchemy](https://bitbucket.org/zzzeek/sqlalchemy/src/c8873b31f0c87ba0d1a7518b36af7151dec34be4/lib/sqlalchemy/engine/base.py?at=master#cl-878) I worked out that I could instead do

```
db = dataset.Database(reflectMetadata=False)

def handler():
    try:
        db.metadata.reflect(db.engine)
        return db['myawesometable'].all()
    except SQLNotConnected:
        return 503
```

But this is very awkward. I would like this step to happen in-line, without me having to think about it.  [I see](https://github.com/pudo/dataset/blob/a797e9035720c24da712fb565a1ce6aa4feeac37/dataset/persistence/database.py#L193) that `update_table()` updates dataset's view of the schema as well; I tried calling `db.update_table(name)`, but I am taking `name` from user input and in that case dataset creates a new, empty table inside of dataset but not in the database: I get 
```
ProgrammingError: (ProgrammingError) (1146, "Table 'username_dbname.landscpa` doesn't exist") 'ALTER TABLE landscpa ADD COLUMN sanche INTEGER' ()
```
if I try to use it (because of the [line](https://github.com/pudo/dataset/blob/a797e9035720c24da712fb565a1ce6aa4feeac37/dataset/persistence/database.py#L197)
```
        self._tables[table_name] = SQLATable(table_name, self.metadata)
```
which assumes `table_name` exists without checking if it actually does first.
).

In short, I want every line inside of dataset that might call out to the backend to be preceeded by a call to `metadata.reflect` to trigger a reconnect attempt and a schema sync. Or at least, I want to be able to make that assumption; I know that transactions might get fiddly about that.

**Comment by pudo** [comment]
Ok, I've thought about this a bit (well, a year) and it seems to me like the use case is too fringe to actually put it in the main code base, since it can easily be addressed by calling ``db.update_table`` before executing further statements in your request cycle. Closing.
